src/lambda_vm/compiler.ml

```diff
@@ -0,0 +1,134 @@
+open Ast
+open Ir
+open Checks
+
+module String_map = Map.Make (String)
+
+type error = (* user program bugs *)
+  | Undefined_variable
+exception Error of error
+let raise error = raise (Error error)
+
+let compile_prim prim =
+  match prim with
+  | Neg -> P_neg
+  | Add -> P_add
+  | Sub -> P_sub
+  | Mul -> P_mul
+  | Div -> P_div
+  | Rem -> P_rem
+  | And -> P_and
+  | Or -> P_or
+  | Xor -> P_xor
+  | Lsl -> P_lsl
+  | Lsr -> P_lsr
+  | Asr -> P_asr
+
+module Vars = Map_with_cardinality.Make (String)
+
+let burn_gas gas vars code =
+  (match code with
+  | Var _
+  | Lam _ ->
+    let cardinality = Vars.cardinal vars in
+    Gas.burn_log2 gas ~cardinality
+  | App _
+  | Const _
+  | Prim _
+  | If _
+  | Pair _
+  | Fst _
+  | Snd _ ->
+    Gas.burn_constant gas);
+  check_gas gas
+
+let rec compile_expr ~stack gas next_ident vars code =
+  let stack = stack - 1 in
+
+  check_stack ~stack;
+  burn_gas gas vars code;
+
+  let compile_expr vars code = compile_expr ~stack gas next_ident vars code in
+
+  match code with
+  | Var var -> (
+    match Vars.find var vars with
+    | Some ident -> E_var ident
+    | None -> raise Undefined_variable)
+  | Lam (var, body) ->
+    let ident =
+      let ident = !next_ident in
+      next_ident := Ident.next ident;
```

Comment:
Why not just write: 
`let next_ident := Ident.next !next_ident`?

Reply:
I don't think that's valid OCaml